### PR TITLE
Fix: Parties

### PR DESCRIPTION
**Pull request by 0x4808D**
- Fixes reconnecting / warping appearing offline
- Fixes party leader turnover logic
- Fixes party members not leaving correctly
- Fixes not being able to kick offline members
- Fixes party packet structure being incorrect in some places

TODO:
- fix regen bug (spams the UpdateHitpoints packet)

#787 

[.](https://www.youtube.com/watch?v=jDBPmEAheCY)

### CODE REVIEW

File: MapleServer2/Managers/PartyManager.cs [hunk 1]
```diff
@@ -16,9 +16,10 @@ public void AddParty(Party party)
         PartyList.Add(party.Id, party);
     }
 
-    public void RemoveParty(Party party)
+    public void RemoveParty(Party? party)
     {
-        PartyList.Remove(party.Id);
+        PartyList.Remove(party?.Id ?? -1);
```
Comment (AlanMorel) [inline MapleServer2/Managers/PartyManager.cs:21]: If the party is null, should we not just return here instead?

File: MapleServer2/Packets/PartyPacket.cs [hunk 2]
```diff
@@ -75,27 +69,36 @@ public static PacketWriter Kick(Player player)
...
         pWriter.WriteBool(joinNotice);
         pWriter.WriteInt(party.Id);
         pWriter.WriteLong(party.Leader.CharacterId);
+
         pWriter.WriteByte((byte) party.Members.Count);
 
         foreach (Player member in party.Members)
         {
-            pWriter.WriteBool(!member.Session?.Connected() ?? false);
+            Console.WriteLine(member.Name + " is connected? " + member.Session?.Connected());
```
Comment (AlanMorel) [inline MapleServer2/Packets/PartyPacket.cs:84]: Can this debug line be removed?

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/PacketHandlers/Game/FieldEnterHandler.cs [hunk 3]
```diff
@@ -83,7 +83,7 @@ public override void Handle(GameSession session, PacketReader packet)
 
         if (player.Party is not null)
         {
-            session.Send(PartyPacket.UpdatePlayer(player));
+            player.Party.BroadcastPacketParty(PartyPacket.UpdatePlayer(player));
         }
```
Comment (Zintixx) [inline MapleServer2/PacketHandlers/Game/FieldEnterHandler.cs:87]: couldn't we just do `player.Party?.BroadcastPacketParty(PartyPacket.UpdatePlayer(player));`

**Review by Zintixx — CHANGES_REQUESTED** [review]

File: MapleServer2/Managers/PartyManager.cs (hunk 1 above)
Comment (0x4808D) [reply MapleServer2/Managers/PartyManager.cs]: Party should never be null here anyways, leftovers.

**Review by Zintixx — APPROVED** [review]

**Review by AlanMorel — APPROVED** [review]

**Pull request merged**